<a href="https://colab.research.google.com/github/LoayEl/ds-dev-fall-2025/blob/loay-elasmar-week-03/Week-03-EDA-and-Dashboards/exercise/movie_rating_dashboard.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [42]:
import pandas as pd
import altair as alt
import plotly.express as px

df= pd.read_csv('data/movie_ratings.csv')

#print(df.info())
print(df.tail())

        user_id  movie_id  rating            timestamp  age gender  \
212590      716       204       5  1997-11-17 19:39:03   36      F   
212591      276      1090       1  1997-09-20 22:49:55   21      M   
212592       13       225       2  1997-12-17 22:52:36   47      M   
212593       13       225       2  1997-12-17 22:52:36   47      M   
212594       12       203       3  1997-11-19 17:13:03   28      F   

           occupation zip_code                      title    year  decade  \
212590  administrator    44265  Back to the Future (1985)  1985.0  1980.0   
212591        student    95064              Sliver (1993)  1993.0  1990.0   
212592       educator    29206      101 Dalmatians (1996)  1996.0  1990.0   
212593       educator    29206      101 Dalmatians (1996)  1996.0  1990.0   
212594          other    06405          Unforgiven (1992)  1992.0  1990.0   

          genres  rating_year  
212590    Sci-Fi         1997  
212591  Thriller         1997  
212592  Children    

In [18]:
df['genres'].value_counts()

genres
Drama          39895
Comedy         29832
Action         25589
Thriller       21872
Romance        19461
Adventure      13753
Sci-Fi         12730
War             9398
Crime           8055
Children        7182
Horror          5317
Mystery         5245
Musical         4954
Animation       3605
Western         1854
Film-Noir       1733
Fantasy         1352
Documentary      758
unknown           10
Name: count, dtype: int64

In [24]:
genre_counts=df['genres'].value_counts().reset_index()
genre_counts.columns=['genre','count']
genre_counts



genre  count
0         Drama  39895
1        Comedy  29832
2        Action  25589
3      Thriller  21872
4       Romance  19461
5     Adventure  13753
6        Sci-Fi  12730
7           War   9398
8         Crime   8055
9      Children   7182
10       Horror   5317
11      Mystery   5245
12      Musical   4954
13    Animation   3605
14      Western   1854
15    Film-Noir   1733
16      Fantasy   1352
17  Documentary    758
18      unknown     10

In [32]:
fig = px.bar(
    genre_counts,
    x='count',
    y='genre',
    orientation='h',
    title="Movie Genres by Ratings",
    labels={'count': 'Number of Ratings', 'genre': 'Genre'}
)
fig.show()

**Most rated Genres:**

-Drama leads with close to 40k ratings and Comedy is also up there with around 30k ratings

**Moderately rated Genres:**

-Action and Thriller with ratings from 21-25k
-Romance, Adventure, and Sci-Fi with ratings from 12k-20

**Least Rated Genres:**

-Documenttary being the lowest reviewed with less than 1k reviews
-Fantasy & Film Noir with a little higher than 1k


In [36]:
genre_rating=df.groupby('genres')['rating'].mean().reset_index()
genre_rating=genre_rating.sort_values(by='rating', ascending=False)
genre_rating

genres    rating
9     Film-Noir  3.921523
16          War  3.815812
7         Drama  3.687379
6   Documentary  3.672823
12      Mystery  3.638132
5         Crime  3.632278
13      Romance  3.621705
17      Western  3.613269
2     Animation  3.576699
14       Sci-Fi  3.560723
11      Musical  3.521397
15     Thriller  3.509007
1     Adventure  3.503527
0        Action  3.480245
4        Comedy  3.394073
3      Children  3.353244
10       Horror  3.290389
8       Fantasy  3.215237
18      unknown  3.200000

In [41]:
fig2= px.bar(
    genre_rating,
    x='rating',
    y='genres',
    orientation='h',
    title= 'Average Rating Per Genre',
    labels={'rating': 'Average Rating', 'genres':'Genre'}
)
fig2.update_layout(
    xaxis=dict(range=[3.1,4.0])
)
fig2.show()

**Highest rated Genres:**

-Film Noir(3.92),War(3.81) have the highest

-while they arent the most popular their fans love its movies very highly

**Lowest Rated Genres:**

-Fantasy (3.21), Horror (3.29), and Children (3.35) have the lowest ratings

In [43]:
df['year'].value_counts()

year
1996.0    37505
1997.0    33760
1995.0    25229
1994.0    18046
1993.0    12778
          ...  
1943.0       45
1936.0       39
1932.0       24
1930.0       18
1926.0        2
Name: count, Length: 71, dtype: int64

In [44]:
year_rating=df.groupby('year')['rating'].mean().reset_index()
year_rating

year    rating
0   1922.0  3.555556
1   1926.0  3.000000
2   1930.0  3.777778
3   1931.0  4.000000
4   1932.0  3.833333
..     ...       ...
66  1994.0  3.426577
67  1995.0  3.443458
68  1996.0  3.312172
69  1997.0  3.325326
70  1998.0  3.323529

[71 rows x 2 columns]

In [48]:
fig3 = px.line(
    year_rating,
    x='year',
    y='rating',
)
fig3.update_xaxes(dtick=10)
fig3.show()

- Ratings were higher in older movies (1930–1960), with peaks like 1942 (4.4)
- From the 1970s forwards, ratings show a downward trend, reaching around 3.3 by the 1990
- Overall, older films seem to be rated more favorably than newer one

In [52]:
#df['title'].value_counts()
movie_stats=df.groupby('title').agg(
    avg_rating=('rating', 'mean'),
    num_ratings=('rating', 'count')
).reset_index()
movie_stats

title  avg_rating  num_ratings
0                 'Til There Was You (1997)    2.333333           18
1                              1-900 (1994)    2.600000            5
2                     101 Dalmatians (1996)    2.908257          218
3                       12 Angry Men (1957)    4.344000          125
4                                187 (1997)    3.024390           41
...                                     ...         ...          ...
1659                   Young Guns II (1990)    2.772727          132
1660  Young Poisoner's Handbook, The (1995)    3.341463           41
1661                Zeus and Roxanne (1997)    2.166667            6
1662                                unknown    3.444444            9
1663     Á köldum klaka (Cold Fever) (1994)    3.000000            2

[1664 rows x 3 columns]

In [55]:
best_50=(movie_stats[movie_stats['num_ratings']>= 50].sort_values(by='avg_rating', ascending=False).head(5))
best_50

title  avg_rating  \
318                               Close Shave, A (1995)    4.491071   
1281                            Schindler's List (1993)    4.466443   
1652                         Wrong Trousers, The (1993)    4.466102   
273                                   Casablanca (1942)    4.456790   
1597  Wallace & Gromit: The Best of Aardman Animatio...    4.447761   

      num_ratings  
318           336  
1281          596  
1652          236  
273           729  
1597           67

In [56]:
best_150=(movie_stats[movie_stats['num_ratings']>= 150].sort_values(by='avg_rating', ascending=False).head(5))
best_150

title  avg_rating  num_ratings
318              Close Shave, A (1995)    4.491071          336
1281           Schindler's List (1993)    4.466443          596
1652        Wrong Trousers, The (1993)    4.466102          236
273                  Casablanca (1942)    4.456790          729
1317  Shawshank Redemption, The (1994)    4.445230          283

In [68]:
fig4= px.bar(
    best_50,
    x='avg_rating',
    y='title',
    orientation='h',
    title= 'Top 5 Movies With At Least 50 Reviews ',
    text='avg_rating',
    labels={'avg_rating': 'Average Rating'}
)

fig4.update_xaxes(range=[4.3, 4.55])
fig4.update_traces(texttemplate='%{text:.2f}', textposition='outside')
fig4.show()

The best-rated movies all have very high averages (4.45–4.49).

A Close Shave”(1995) leads with the highest rating (4.49).

In [71]:
fig4= px.bar(
    best_150,
    x='avg_rating',
    y='title',
    orientation='h',
    title= 'Top 5 Movies With At Least 150 Reviews ',
    text='avg_rating',
    labels={'avg_rating': 'Average Rating'}
)

fig4.update_xaxes(range=[4.3, 4.55])
fig4.update_traces(texttemplate='%{text:.2f}', textposition='outside')
fig4.show()

- When we raise the bar to 150 reviews Shawshank Redemption takes a spot in the top 5

- This pushes out Wallace & Gromit, which was in the top 5 with the 50 review limit

- The other four movies stay the same, showing they are strong no matter the cutoff